### PR TITLE
Add FlushAll [cherry-pick#22942]

### PR DESCRIPTION
**Pull request by bigsheeper**
issues: https://github.com/milvus-io/milvus/issues/22921

### CODE REVIEW

File: internal/proxy/impl.go [hunk 1]
```diff
@@ -3484,6 +3485,61 @@ func (node *Proxy) CalcDistance(ctx context.Context, request *milvuspb.CalcDista
...
+		log.Warn("FlushAll failed", zap.String("err", showColRsp.GetStatus().GetReason()))
+		resp.Status.Reason = showColRsp.GetStatus().GetReason()
+		return resp, nil
+	}
+	flushRsp, err := node.Flush(ctx, &milvuspb.FlushRequest{
+		Base:            commonpbutil.NewMsgBase(commonpbutil.WithMsgType(commonpb.MsgType_Flush)),
+		CollectionNames: showColRsp.GetCollectionNames(),
+	})
+	if err != nil {
+		return nil, err
```
Comment (longjiquan) [inline internal/proxy/impl.go:3519]: Ditto, same above.

File: internal/proxy/impl.go [hunk 2]
```diff
@@ -3484,6 +3485,61 @@ func (node *Proxy) CalcDistance(ctx context.Context, request *milvuspb.CalcDista
...
+	}
+	if flushRsp.GetStatus().GetErrorCode() != commonpb.ErrorCode_Success {
+		log.Warn("FlushAll failed", zap.String("err", flushRsp.GetStatus().GetReason()))
```
Comment (longjiquan) [inline internal/proxy/impl.go:3522]: Ditto, same above.

File: internal/proxy/impl.go [hunk 3]
```diff
@@ -3484,6 +3485,61 @@ func (node *Proxy) CalcDistance(ctx context.Context, request *milvuspb.CalcDista
...
+		resp.Status.Reason = flushRsp.GetStatus().GetReason()
+		return resp, nil
+	}
+
+	// allocate current ts as FlushAllTs
+	ts, err := node.tsoAllocator.AllocOne(ctx)
+	if err != nil {
+		log.Warn("FlushAll failed", zap.Error(err))
```
Comment (longjiquan) [inline internal/proxy/impl.go:3530]: Use `info` here.

File: internal/proxy/impl.go [hunk 4]
```diff
@@ -4010,6 +4066,32 @@ func (node *Proxy) GetFlushState(ctx context.Context, req *milvuspb.GetFlushStat
...
+	resp := &milvuspb.GetFlushAllStateResponse{}
+	if !node.checkHealthy() {
+		resp.Status = unhealthyStatus()
+		log.Warn("GetFlushAllState failed, closed server")
+		return resp, nil
+	}
+
+	resp, err = node.dataCoord.GetFlushAllState(ctx, req)
+	if err != nil {
+		log.Warn("GetFlushAllState failed", zap.Error(err))
```
Comment (longjiquan) [inline internal/proxy/impl.go:4088]: Ditto, same above.

**Review by longjiquan — CHANGES_REQUESTED** [review]

File: internal/datacoord/services.go [hunk 5]
```diff
@@ -1180,6 +1180,51 @@ func (s *Server) GetFlushState(ctx context.Context, req *milvuspb.GetFlushStateR
...
+		setNotServingStatus(resp.Status, s.GetStateCode())
+		return resp, nil
+	}
+
+	showColRsp, err := s.rootCoordClient.ShowCollections(ctx, &milvuspb.ShowCollectionsRequest{
+		Base: commonpbutil.NewMsgBase(
+			commonpbutil.WithMsgType(commonpb.MsgType_ShowCollections),
+		)})
+	if err = VerifyResponse(showColRsp, err); err != nil {
+		log.Warn("failed to ShowCollections", zap.Error(err))
```
[the remark this answers is not in the document]
Comment (bigsheeper) [reply internal/datacoord/services.go:1197]: Why use Info here？
`ShowCollections ` is better to search log when debugging, any strong reason to use `show collections`?

File: internal/datacoord/services.go [hunk 6]
```diff
@@ -1180,6 +1180,51 @@ func (s *Server) GetFlushState(ctx context.Context, req *milvuspb.GetFlushStateR
...
+			),
+			CollectionID: collection,
+		})
+		if err = VerifyResponse(describeColRsp, err); err != nil {
+			log.Warn("failed to DescribeCollectionInternal", zap.Error(err))
+			resp.Status.Reason = err.Error()
+			return resp, nil
+		}
+		for _, channel := range describeColRsp.GetVirtualChannelNames() {
+			channelCP := s.meta.GetChannelCheckpoint(channel)
```
[the remark this answers is not in the document]
Comment (bigsheeper) [reply internal/datacoord/services.go:1215]: It's ok, just wait for next round to GetFlushAllState.

File: internal/datacoord/services.go [hunk 7]
```diff
@@ -1180,6 +1180,51 @@ func (s *Server) GetFlushState(ctx context.Context, req *milvuspb.GetFlushStateR
...
+	for _, collection := range showColRsp.GetCollectionIds() {
+		describeColRsp, err := s.rootCoordClient.DescribeCollectionInternal(ctx, &milvuspb.DescribeCollectionRequest{
+			Base: commonpbutil.NewMsgBase(
+				commonpbutil.WithMsgType(commonpb.MsgType_DescribeCollection),
+			),
+			CollectionID: collection,
+		})
+		if err = VerifyResponse(describeColRsp, err); err != nil {
+			log.Warn("failed to DescribeCollectionInternal", zap.Error(err))
+			resp.Status.Reason = err.Error()
```
Comment (jaime0815) [inline internal/datacoord/services.go:1211]: should set  error code of describe the response to this status

File: internal/datacoord/services.go [hunk 8]
```diff
@@ -1180,6 +1180,51 @@ func (s *Server) GetFlushState(ctx context.Context, req *milvuspb.GetFlushStateR
...
+		return resp, nil
+	}
+
+	showColRsp, err := s.rootCoordClient.ShowCollections(ctx, &milvuspb.ShowCollectionsRequest{
+		Base: commonpbutil.NewMsgBase(
+			commonpbutil.WithMsgType(commonpb.MsgType_ShowCollections),
+		)})
+	if err = VerifyResponse(showColRsp, err); err != nil {
+		log.Warn("failed to ShowCollections", zap.Error(err))
+		resp.Status.Reason = err.Error()
```
Comment (jaime0815) [inline internal/datacoord/services.go:1198]: should set error code of describe the response to this status

**Pull request merged**